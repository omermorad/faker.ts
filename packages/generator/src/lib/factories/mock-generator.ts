import { Class, Faker } from '@mockinbird/types';
import { ClassProcessor } from '@mockinbird/parser';
import { ClassReflector } from '@mockinbird/reflect';
import { MockDecoratorFactoryOptions } from '../../types/mock-decorator-factory-options.interface';

export class MockGenerator {
  /**
   * Return an object with all the properties decorated by the 'Mock' Decorator
   *
   * @example
   * class Person { @Mock() name: string }
   * MockFactory.create(Person) will return an object { name: <random-string> }
   *
   * @param target
   */
  public static create<TClass = any>(target: Class<TClass>): TClass;

  /**
   * Return an array of objects with all the properties decorated by the
   * 'Mock' Decorator (if 'count' is greater than 1)
   *
   * @example
   * class Person { @Mock() name: string }
   * MockFactory.create(Person, { count: 3, locale: 'es' }) will return an
   * array of objects [{ name: <random-string> }, { name: <random-string> },
   * { name: <random-string> }]
   *
   * Passing a 'locale' property will set a different locale for faker calls
   * The default locale is 'en' (english)
   *
   * @param target
   * @param options
   */
  public static create<TClass = any>(target: Class<TClass>, options: MockDecoratorFactoryOptions): TClass[];

  /**
   * Return one or many objects (array) with all the properties decorated
   * by the Mock decorator
   *
   * @param targetClass
   * @param options
   */
  public static create<TClass = any>(
    targetClass: Class<TClass>,
    options?: MockDecoratorFactoryOptions
  ): TClass | TClass[] {
    const { count = 1, locale = ClassProcessor.DEFAULT_LOCALE } = options || {};
    const processor = new ClassProcessor<TClass>(Faker, new ClassReflector(), locale);

    if (!count || count === 1) {
      return processor.process(targetClass);
    }

    const classInstances: TClass[] = [];

    for (let i = 1; i <= count; i++) {
      classInstances.push(processor.process(targetClass));
    }

    return classInstances;
  }
}
