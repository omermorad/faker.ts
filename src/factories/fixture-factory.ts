import faker from 'faker';
import { FixturesCreator } from '../handlers/fixtures-creator';
import { ClassLiteral, ClassType } from '../types/class.type';
import { GeneratorFactoryOptions } from '../types/generator-factory-options.interface';

export class FixtureFactory {
  public static create<T = unknown>(target: ClassType<T>, options: GeneratorFactoryOptions): ClassLiteral<T>[];

  public static create<T = unknown>(target: ClassType<T>): ClassLiteral<T>;

  public static create<T = unknown>(
    target: ClassType<T>,
    options?: GeneratorFactoryOptions
  ): ClassLiteral<T> | ClassLiteral<T>[] {
    const { total = 1, locale = FixturesCreator.DEFAULT_LOCALE } = options || {};

    const factory = new FixturesCreator<T>(faker, locale);

    if (!total || total === 1) {
      return factory.createForTarget(target) as ClassLiteral<T>;
    }

    const objects: ClassLiteral<T>[] = []; // test

    for (let i = 1; i <= total; i++) {
      objects.push(factory.createForTarget(target));
    }

    return objects;
  }
}
