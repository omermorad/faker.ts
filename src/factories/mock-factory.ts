import faker from 'faker';
import { ClassProcessor } from '../class-processor';
import { ClassLiteral, Class } from '../types/mock-options.type';
import { MockFactoryOptions } from '../types';
import { ClassReflector } from '../class-reflector';

export class MockFactory {
  /**
   * Return an object with all the properties decorated by the 'Mock' Decorator
   *
   * @example
   * class Person { @Mock() name: string }
   * MockFactory.create(Person) will return an object { name: <random-string> }
   *
   * @param target
   */
  public static create<T = unknown>(target: Class<T>): ClassLiteral<T>;

  /**
   * Return an array of objects with all the properties decorated by the
   * 'Mock' Decorator (if 'count' is greater than 1)
   *
   * @example
   * class Person { @Mock() name: string }
   * MockFactory.create(Person, { count: 3, locale: 'es' }) will return an
   * array of objects [{ name: <random-string> }, { name: <random-string> },
   * { name: <random-string> }]
   *
   * Passing a 'locale' property will set a different locale for faker calls
   * The default locale is 'en' (english)
   *
   * @param target
   * @param options
   */
  public static create<T = unknown>(target: Class<T>, options: MockFactoryOptions): ClassLiteral<T>[];

  /**
   * Return one or many objects (array) with all the properties decorated
   * by the Mock decorator
   *
   * @param target
   * @param options
   */
  public static create<T = unknown>(
    target: Class<T>,
    options?: MockFactoryOptions
  ): ClassLiteral<T> | ClassLiteral<T>[] {
    const { count = 1, locale = ClassProcessor.DEFAULT_LOCALE } = options || {};

    const factory = new ClassProcessor<T>(faker, new ClassReflector(), locale);

    if (!count || count === 1) {
      return factory.process(target) as ClassLiteral<T>;
    }

    const objects: ClassLiteral<T>[] = [];

    for (let i = 1; i <= count; i++) {
      objects.push(factory.process(target));
    }

    return objects;
  }
}
