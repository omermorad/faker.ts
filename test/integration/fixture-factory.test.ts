import { Fixture } from '../../src/decorators/fixture.decorator';
import { FixtureFactory } from '../../src/factories/fixture-factory';

enum Activity {
  VeryActive,
  Active,
  NotActive,
}

describe('Fixture Factory - Snapshot test', () => {
  describe('Absolute values', () => {
    it('should use the actual strait value from the decorator', () => {
      class Dummy {
        @Fixture('Johnny Boy')
        name: string;

        @Fixture(1234)
        num: number;

        @Fixture(true)
        binary: boolean;

        @Fixture(new Date('01/01/1970'))
        date: Date;
      }

      const result = FixtureFactory.create(Dummy);
      expect(result).toMatchSnapshot();
    });
  });
});
