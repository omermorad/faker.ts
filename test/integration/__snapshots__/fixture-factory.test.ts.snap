// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fixture Factory - Snapshot test Absolute values should use the actual strait value from the decorator 1`] = `
Object {
  "binary": true,
  "date": "Mon Jan 18 2021 23:43:52 GMT+0200 (Israel Standard Time)",
  "name": "Johnny Boy",
  "num": 1234,
}
`;
